cpp/src/arrow/compute/kernels/random.cc

```diff
@@ -0,0 +1,55 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/util/random.h"
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/kernels/common.h"
+#include "arrow/compute/registry.h"
+
+#include <random>
+
+namespace arrow {
+namespace compute {
+namespace internal {
+namespace {
+
+Status ExecRandom(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  float value = random::generate_canonical<float>(gen);
+  BoxScalar<FloatType>::Box(value, out->scalar().get());
+  return Status::OK();
```

Comment:
I know it's still a draft, but it's better to agree on some requirements early.
Which RNG to use? I prefer PCG which is already vendored [1].
Shall we use our own `random::generate_canonical` which is to workaround performance issue on Arm [2], or functions provided by STL, or something else? I'm not sure.
@pitrou 

[1] https://github.com/apache/arrow/tree/master/cpp/src/arrow/vendored/pcg
[2] https://issues.apache.org/jira/browse/ARROW-12533

Reply:
@cyb70289 I'm aware of the history of it, that's why I made it a draft. Just wanted to have the structure in place and trigger CI.

I've tried PCG but I maybe I'm holding it wrong since it only generates 0, haven't really looked into it. I just used `std::mt19937` to get this going, which kinda defeats the purpose of providing our own random algorithm anyway.

I'd go with STL, yes. It's not clear to me from previous discussion why we're reluctant to just go with that.

Reply:
Just checked and PCG always generates a 0 first regardless of the seed. I'm not sure this behavior is acceptable.

Reply:
@asuhan That sounds unexpected.

Reply:
Out of curiosity, I added the following test which passes. So I think you may be mistaken.
```diff
diff --git a/cpp/src/arrow/testing/random_test.cc b/cpp/src/arrow/testing/random_test.cc
index 6b20d821d..31a7f95b4 100644
--- a/cpp/src/arrow/testing/random_test.cc
+++ b/cpp/src/arrow/testing/random_test.cc
@@ -14,6 +14,9 @@
 // KIND, either express or implied.  See the License for the
 // specific language governing permissions and limitations
 // under the License.
+
+#include <random>
+
 #include <gtest/gtest.h>
 
 #include "arrow/array.h"
@@ -516,5 +519,13 @@ TEST(BernoulliTest, Basic) {
   }
 }
 
+TEST(PcgRandom, Basic) {
+  std::random_device device;
+  pcg64 gen(device());
+
+  ASSERT_NE(gen(), 0);
+  ASSERT_NE(gen(), 0);
+}
+
 }  // namespace random
 }  // namespace arrow
```

Reply:
By the way, the 64-bit variants of PCG will probably be significantly faster (because you only need one number generation instead of two to generate a 64-bit float).

Reply:
Should we make `random()` return double? I made it return float.

Also, just checked `pcg32_fast` again and indeed it returns 0 for all seeds I've tried (numbers <= 1000), but `pcg32` doesn't show that behavior (neither do `pcg64` nor `pcg64_fast`).

Reply:
Yes, it should return `double`. I think we should simply use `pcg64` here.

As for the `pcg32_fast` issue, perhaps you can try reproducing with the upstream library (https://github.com/imneme/pcg-cpp) and report a bug there?